api/src/main/java/com/cloud/host/Status.java

```diff
@@ -131,6 +131,7 @@ public Status getNextStatus(Event e) throws NoTransitionException {
         s_fsm.addTransition(Status.Up, Event.PingTimeout, Status.Alert);
         s_fsm.addTransition(Status.Up, Event.AgentDisconnected, Status.Alert);
         s_fsm.addTransition(Status.Up, Event.ShutdownRequested, Status.Disconnected);
+        s_fsm.addTransition(Status.Disconnected, Event.ShutdownRequested, Status.Disconnected);
```

Comment:
do we need a transition from disconnected to disconnected?

Reply:
@DaanHoogland During boot-up of the systemVM , there is a state transition that generally happens which moves the agent state from 'Up' to 'Disconnected', now that we have set the state to disconnected at the stopping phase, this leads to an unknown state transition(during Starting of the VM) as seen in:
```
021-05-10 11:52:57,458 DEBUG [c.c.a.m.AgentManagerImpl] (AgentTaskPool-3:ctx-e4561745) (logid:5a050056) Cannot find next status for ShutdownRequested as current status is Disconnected for agent 5
2021-05-10 11:52:57,458 ERROR [c.c.a.m.AgentManagerImpl] (AgentTaskPool-3:ctx-e4561745) (logid:5a050056) Exception caught while handling disconnect:
com.cloud.utils.exception.CloudRuntimeException: Cannot find next status for ShutdownRequested as current status is Disconnected for agent 5
        at com.cloud.agent.manager.AgentManagerImpl.handleDisconnectWithoutInvestigation(AgentManagerImpl.java:801)
        at com.cloud.agent.manager.ClusteredAgentManagerImpl.handleDisconnect(ClusteredAgentManagerImpl.java:315)
        at com.cloud.agent.manager.ClusteredAgentManagerImpl.handleDisconnectWithoutInvestigation(ClusteredAgentManagerImpl.java:304)
        at com.cloud.agent.manager.AgentManagerImpl$DisconnectTask.runInContext(AgentManagerImpl.java:929)
        at org.apache.cloudstack.managed.context.ManagedContextRunnable$1.run(ManagedContextRunnable.java:48)
        at org.apache.cloudstack.managed.context.impl.DefaultManagedContext$1.call(DefaultManagedContext.java:55)
        at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.callWithContext(DefaultManagedContext.java:102)
        at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.runWithContext(DefaultManagedContext.java:52)
        at org.apache.cloudstack.managed.context.ManagedContextRunnable.run(ManagedContextRunnable.java:45)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)

```
Hence, the particular state transition has been added

Reply:
@Pearl1594 should more transition be added, similar as Event.AgentDisconnected ?

Reply:
@weizhouapache are you referring to the following transitions?
```
s_fsm.addTransition(Status.Disconnected, Event.AgentDisconnected, Status.Disconnected);
s_fsm.addTransition(Status.Down, Event.AgentDisconnected, Status.Down);
s_fsm.addTransition(Status.Alert, Event.AgentDisconnected, Status.Alert);
...

```

Reply:
> @weizhouapache are you referring to the following transitions?
> 
> ```
> s_fsm.addTransition(Status.Disconnected, Event.AgentDisconnected, Status.Disconnected);
> s_fsm.addTransition(Status.Down, Event.AgentDisconnected, Status.Down);
> s_fsm.addTransition(Status.Alert, Event.AgentDisconnected, Status.Alert);
> ...
> ```

@Pearl1594 yeah. since we do not know what the original state is , it might be better to add transitions for all possible state.
```
# grep Event.AgentDisconnected ./api/src/main/java/com/cloud/host/Status.java
        s_fsm.addTransition(Status.Connecting, Event.AgentDisconnected, Status.Alert);
        s_fsm.addTransition(Status.Up, Event.AgentDisconnected, Status.Alert);
        s_fsm.addTransition(Status.Disconnected, Event.AgentDisconnected, Status.Disconnected);
        s_fsm.addTransition(Status.Down, Event.AgentDisconnected, Status.Down);
        s_fsm.addTransition(Status.Alert, Event.AgentDisconnected, Status.Alert);
        s_fsm.addTransition(Status.Rebalancing, Event.AgentDisconnected, Status.Rebalancing);
```